File: Problem131_PalindromePartitioning.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    vector<vector<string>> res;
    vector<string> temp;
    bool isPalindrome(string s, int l, int r) {
        while (l <= r) {
            if (s[l]!=s[r]) return false;
            l++;
            r--;
        }
        return true;
    }
    void backtrack(int start, string s) {
        int n=s.length();
        if (start==n) {
            res.push_back(temp);
            return;
        }
        for (int i=start;i<n;i++) {
            if (isPalindrome(s, start, i)) {
                temp.push_back(s.substr(start, i-start+1));
                backtrack(i+1, s);
                temp.pop_back();
            }
        }
    }
    vector<vector<string>> partition(string s) {
        backtrack(0, s);
        return res;
    }
};

int main() {
    Solution solve;
    string s = "aaab";
    vector<vector<string>> res = solve.partition(s);
    for (auto& row : res) {
        for (string s : row) cout << s << " ";
        cout << endl;
    }
}

File: Problem131_PalindromePartitioning.py
from typing import List

class Solution:
    def partition(self, s: str) -> List[List[str]]:
        res = []
        oneRes = []
        n = len(s)
        def isPalindrome(s, l, r):
                while l<=r:
                    if s[l]!=s[r]: return False
                    l+=1
                    r-=1
                return True
        def backtrack(start, s):
            if start == n:
                res.append(oneRes.copy())
                return
            for i in range(start, n):
                if isPalindrome(s, start, i):
                    oneRes.append(s[start:i+1])
                    backtrack(i+1, s)
                    oneRes.pop()
        backtrack(0, s)
        return res

solve = Solution()
s = "aab"
print(solve.partition(s))